build-tools/packages/build-cli/src/BasePackageCommand.ts

```diff
@@ -127,31 +173,70 @@ export abstract class PackageCommand<
 		}
 	}
 
+	/**
+	 * Runs processPackage for each package in a release group, exlcuding the root.
+	 */
+	private async processReleaseGroup(
```

Comment:
I think separating the concerns of finding the right packages, and running an operation on them would be cleaner. Having this return a list of packages, then having the user of this add it to the full set its going to process would accomplish this.

This approach would also make actions like detecting duplicates, or applying multiple different operations to a set of packages easier.

It would also allow parallelizing across multiple release groups while having a single scheduler to avoid having too many in progress tasks and mixed status reporting.

Reply:
Also making async methods have useful return values tends to help reduce the risk or unawaited promises or unexpected concurrency.